Source/msg.cpp

```diff
@@ -707,7 +707,18 @@ DWORD OnRequestGetItem(const TCmd *pCmd, Player &player)
 	if (gbBufferMsgs != 1 && IOwnLevel(player.plrlevel) && IsGItemValid(message)) {
 		const Point position { message.x, message.y };
 		if (GetItemRecord(message.dwSeed, message.wCI, message.wIndx)) {
-			int ii = FindGetItem(message.wIndx, message.wCI, message.dwSeed);
+			int ii = -1;
+			if (InDungeonBounds(position)) {
+				ii = abs(dItem[position.x][position.y]) - 1;
+				if (ii >= 0 && !Items[ii].KeyAttributesMatch(message.dwSeed, static_cast<_item_indexes>(message.wIndx), message.wCI)) {
+					ii = -1;
+				}
+			}
+
+			if (ii == -1) { // No item at the target position or the key attributes don't match, so try find a matching item.
+				ii = FindGetItem(message.wIndx, message.wCI, message.dwSeed);

```

Comment:
Is there a valid case where this could happen?

Reply:
In the current code, its more likely if a remote player in a multiplayer game is dropping items on a different floor to the local player due to the way item deltas are recorded and processed. #2932 updates the code that "plays back" the history of actions to handle dropped items a bit better which should hopefully make this check unnecessary.

Potentially could still be an issue with lag though, but in that case it might be safe to just fail to pick up an item and let the game stay desynced.

Reply:
Damn you have so many clean ups going, I'm having a hard time keeping up :D